New Coding Guide - Audit Logs

ChatDefender opened the pull request:
I thought it'd be nice to create a small and understandable guide to audit logs.

ChatDefender committed:
Create AuditLogs.md

ChatDefender committed:
Rename AuditLogs.md to understanding-and-using-audit-logs.md

YorkAARGH commented on coding-guides/understanding-and-using-audit-logs.md:
```diff
@@ -0,0 +1,142 @@
+let it be viewing them for certain actions to adding things into the audit logs. 
+This guide will explain everything about audit logs and how to use them.
+
+The first thing that you will need is a working Discord Bot. If you do not have one, please visit [Your First Bot](./getting-started/your-basic-bot.md) to get started.
```
you don't need to add the `.` to the file url

YorkAARGH commented on coding-guides/understanding-and-using-audit-logs.md:
```diff
@@ -0,0 +1,142 @@
+The first thing that you will need is a working Discord Bot. If you do not have one, please visit [Your First Bot](./getting-started/your-basic-bot.md) to get started.
+
+Now, some things to take note of. The bot will need some permissions. The main permission the bot will need
+is 'VIEW_AUDIT_LOGs'. This permission allows the bot to view the audit logs and do whatever you want to, obeying the Discord ToS that is. 
```
please use back ticks for `VIEW_AUDIT_LOGS`, makes the text stand out, also you accidentally lower cased the `s` on `LOGS`

YorkAARGH commented on coding-guides/understanding-and-using-audit-logs.md:
```diff
@@ -0,0 +1,142 @@
+  // Firstly, we need a logs channel. 
+  const logs = message.guild.channels.find('name', 'logs');
+  
+  // If there is no logs channel, we can create it if we have the 'MANAGE_CHANNELS' permission
```
This is a nice little touch, please mention that this is completely optional.

ChatDefender committed:
Update SUMMARY.md

ChatDefender committed:
Update SUMMARY.md

YorkAARGH commented on coding-guides/understanding-and-using-audit-logs.md:
```diff
@@ -0,0 +1,142 @@
+  
+  // If we do not have permissions, console log both errors
+  if (!message.guild.me.hasPermission('MANAGE_CHANNELS') && !logs) { 
+    console.log('The logs channel does not exist and tried to create the channel but am lacking permissions')
```
Grammer mistake here, should be `I am` or `I'm`

YorkAARGH commented on SUMMARY.md:
```diff
@@ -23,6 +23,7 @@
   * [Cleverbot Integration](coding-guides/cleverbot-integration.md)
   * [Selfbots, the greatest thing in the universe](examples/selfbots-are-awesome.md)
   * [Using Emojis](coding-guides/using-emojis.md)
+  * [Using and Understanding Audit Logs](coding-guides/using-and-understanding-audit-logs.md)
```
Change it to `Using Audit Logs` due to how long the link name would be.

ChatDefender committed:
Update SUMMARY.md

ChatDefender committed:
Update understanding-and-using-audit-logs.md

YorkAARGH commented on coding-guides/understanding-and-using-audit-logs.md:
```diff
@@ -0,0 +1,142 @@
+  targetType: 'MESSAGE',
+  actionType: 'DELETE',
+  action: 'MESSAGE_DELETE',
+  reason: null,
```
Maybe mention that some actions can be given a reason, such as kick / ban

YorkAARGH reviewed (changes requested):
This is pretty well written, also make sure you mention _which_ version of discord.js this is for, as some people still think it's interchangeable between versions.

ChatDefender committed:
Updated Spelling/Gramtical Errors

ChatDefender committed:
Update understanding-and-using-audit-logs.md

ChatDefender committed:
Update understanding-and-using-audit-logs.md

Adjusted spelling errors and corrected link(s)

ChatDefender committed:
Update SUMMARY.md

ChatDefender committed:
Rename understanding-and-using-audit-logs.md to using-audit-logs.md

ghost commented:
Should I add some pictures of what the audit logs looks like, for the reason part specifically?

YorkAARGH commented on coding-guides/using-audit-logs.md:
```diff
@@ -6,8 +6,7 @@ This guide will explain everything about audit logs and how to use them.
 
-Now, some things to take note of. The bot will need some permissions. The main permission the bot will need
-is 'VIEW_AUDIT_LOGs'. This permission allows the bot to view the audit logs and do whatever you want to, obeying the Discord ToS that is. 
+Now, some things to take note of. This guide is using discord.js@11.3.2 The bot will need some permissions. The main permission the bot will need is `'VIEW_AUDIT_LOGS'`. This permission allows the bot to view the audit logs and do whatever you want to, obeying the Discord ToS that is. 
```
_This guide is using discord.js@11.3.2 **T**he bot_, that `T` needs to follow a period, `.` or be lower case `t`

YorkAARGH reviewed (changes requested):
One minor grammatical error

eslachance commented on coding-guides/using-audit-logs.md:
```diff
@@ -0,0 +1,142 @@
+# Understanding Audit Logs
+
+From time to time, users in "An Idiot's Guide Official Server" need to reference audit logs for whatever reason;
```
Why is there a ; here? Probably it should be a comma? Who uses semicolons in english anymore?

eslachance commented on coding-guides/using-audit-logs.md:
```diff
@@ -0,0 +1,142 @@
+
+The first thing that you will need is a working Discord Bot. If you do not have one, please visit [Your First Bot](getting-started/your-basic-bot.md) to get started.
+
+Now, some things to take note of. This guide is using discord.js@11.3.2 The bot will need some permissions. The main permission the bot will need is `'VIEW_AUDIT_LOGS'`. This permission allows the bot to view the audit logs and do whatever you want to, obeying the Discord ToS that is. 
```
I feel like "do whatever you want, obeing the ToS" offers no significant information for the reader. This sentence can end at `view the audit logs.`

eslachance commented on coding-guides/using-audit-logs.md:
```diff
@@ -0,0 +1,142 @@
+Now that the permission has been established. Lets get started!
+
+Firstly, we need to know what we are doing with the audit logs. 
+Lets log who deleted a message using the messageDeleted event. This event will fire whenever a cached message is deleted in a server.
```
Let's*

eslachance commented on coding-guides/using-audit-logs.md:
````diff
@@ -0,0 +1,142 @@
+Lets log who deleted a message using the messageDeleted event. This event will fire whenever a cached message is deleted in a server.
+
+```js
+client.on('messageDeleted', async (message) => {
````
The event is `messageDelete` not `messageDeleted`

eslachance commented on coding-guides/using-audit-logs.md:
```diff
@@ -0,0 +1,142 @@
+  
+  // Now that we have the logs channel created...hopefully...we can send messages to it
+  // before we do that, lets establish who deleted the message
+  // The "type" is how you will be searching through the audit logs. Like roles updated or members banned.
```
Here, you're forgetting one thing. If the original author of the message deletes it, you'll get a messageDelete event, but you won't get an auditLog event because author-deleted messages don't cause a log to appear.

eslachance commented on coding-guides/using-audit-logs.md:
````diff
@@ -0,0 +1,142 @@
+So, our `executor` should be look similar to this:
+
+```js
+  const executor = await message.guild.fetchAuditLogs({type: 'MESSAGE_DELETE'}).then(audit => audit.entries.first().executor.username)
````
Generally speaking, people will actually want to use more than just the username of an audit log. It's best practice to actually return the whole entry and then take a property *from* it. that would make more sense.

eslachance commented on coding-guides/using-audit-logs.md:
````diff
@@ -0,0 +1,142 @@
+```js
+  // I always format my messages in strings as it's easier (for me) to understand what I am doing as some strings can get pretty long,
+  // like the $5 foot long from subway. Man I miss that. 
+  const deletedMessageInformation = "A message was deleted in " + message.channel.name + " by " + executor;
````
please learn template literals, and use them. We don't do string concatenation in this guide.

eslachance reviewed (changes requested):
There's a few things that need to be fixed. Otherwise, this is a great addition to the guide!

ChatDefender committed:
Update using-audit-logs.md

"Generally speaking, people will actually want to use more than just the username of an audit log. It's best practice to actually return the whole entry and then take a property from it. that would make more sense." I've tried several things, setting the first entry to a variable using let, created an async function to return the first entry, no go. Any ideas as of how I could do this? It is a great idea.

ChatDefender committed:
Update using-audit-logs.md

Added a few checks to narrow down who deleted the message, whether it was the guy in audit logs or the author.

YorkAARGH commented on coding-guides/using-audit-logs.md:
```diff
@@ -0,0 +1,169 @@
+
+What about bots? Bot messages can be deleted, and they can delete thier own. How can we tell if it is a bot? Great question. The answer is you don't, at least not without some sort of storage data within your own bot. I used a small enmap database that allows me to show that the user was the bot based on where the bot is deleting the messages, whether it was a innapropriate word or a link. 
+
+Let's take a break to explain exactly whats going on in the above code block. The `Date.now()` is getting the current time (in milliseconds). We want to take away 5 seconds for the potential delay in the audit logs. The executor will be retrieving the very latest audit log entry and all of its information that goes with it. What does this information contain? Everything we need for logging. 
```
Um... what code above?, do you mean `below` ?

ChatDefender committed:
Update using-audit-logs.md

eslachance commented on coding-guides/using-audit-logs.md:
```diff
@@ -0,0 +1,169 @@
+
+What about bots? Bot messages can be deleted, and they can delete thier own. How can we tell if it is a bot? Great question. The answer is you don't, at least not without some sort of storage data within your own bot. I used a small enmap database that allows me to show that the user was the bot based on where the bot is deleting the messages, whether it was a innapropriate word or a link. 
+
+Let's take a break to explain exactly whats going on in the below code block. The `Date.now()` is getting the current time (in milliseconds). We want to take away 5 seconds for the potential delay in the audit logs. The executor will be retrieving the very latest audit log entry and all of its information that goes with it. What does this information contain? Everything we need for logging. 
```
I think you mean "Above"

YorkAARGH replied:
When I mentioned it, the code block he was referencing, was below.

eslachance commented on coding-guides/using-audit-logs.md:
````diff
@@ -0,0 +1,169 @@
+})
+```
+
+What about bots? Bot messages can be deleted, and they can delete thier own. How can we tell if it is a bot? Great question. The answer is you don't, at least not without some sort of storage data within your own bot. I used a small enmap database that allows me to show that the user was the bot based on where the bot is deleting the messages, whether it was a innapropriate word or a link. 
````
Not sure what you're trying to say with this paragraph. Whether it's a bot or not doesn't really change what this particular script does, so it's sort of irrelevant. And you saving what users are a bot is weird, since `user.bot` returns a boolean that tells you this (so executor.bot would give that info).

eslachance commented on coding-guides/using-audit-logs.md:
```diff
@@ -0,0 +1,169 @@
+  // Please keep in mind: Discord's audit logs will not log the information if the author of that message deleted it.
+  // I did this with a series of checks. 
+  // We are checking if the channel ID is the same as the id the message is deleted it
+  if ((await message.guild.fetchAuditLogs({type: 'MESSAGE_DELETE'}).then(g => g.entries.first().extra.channel.id === message.channel.id)) 
```
Congrats, in this line and the next 9 line you've just made 7 absolutely useless requests to the API. Fetch the log *once* at the top of the script and then refer to that variable in all other instances. This is what I was talking about with "don't return the executor, return the entry and get its properties". Doing multiple requests is spammy, might break ratelimits, and will slow down this process to hell. 

```js
const entry = await message.guild.fetchAuditLogs({type: 'MESSAGE_DELETE'}).then(g => g.entries.first());
```
Boom, done, now `entry` has all the things you need.

eslachance commented on coding-guides/using-audit-logs.md:
```diff
@@ -0,0 +1,169 @@
+  // We are comparing time as audit logs are sometimes slow. 
+  && (await message.guild.fetchAuditLogs({type: 'MESSAGE_DELETE'}).then(g => g.entries.first().createdTimestamp > (Date.now() - 5000)))) {
+    user = executor
+  } else if ((await message.guild.fetchAuditLogs({type: 'MESSAGE_DELETE'}).then(g => g.entries.first().extra.channel.id === message.channel.id)) 
```
You really don't need 4 different conditions to check if the message is the right one. `entry.message.id === message.id` will tell you if it's the same message, there's no need for any thing else.

eslachance commented on coding-guides/using-audit-logs.md:
````diff
@@ -0,0 +1,169 @@
+```js
+  // I always format my messages in strings as it's easier (for me) to understand what I am doing as some strings can get pretty long,
+  // like the $5 foot long from subway. Man I miss that. 
+  const deletedMessageInformation = `A message was deleted in ${message.channel.name} by ${user}`;
````
Creating a variable for exactly one use is a waste of cycles. 

Do this instead:
```js
logs.send(`A message was deleted in ${message.channel.name} by ${user}`);
```

ChatDefender committed:
Update using-audit-logs.md

ChatDefender committed:
Update SUMMARY.md

ChatDefender committed:
Update SUMMARY.md

ChatDefender committed:
Update SUMMARY.md

ChatDefender committed:
Update SUMMARY.md

ChatDefender committed:
Update SUMMARY.md

ChatDefender committed:
Merge branch 'master' into master

ChatDefender committed:
Update SUMMARY.md

YorkAARGH committed:
Merge branch 'master' into master

YorkAARGH committed:
Merge branch 'master' into master
